### DISCUSSION TITLE
Query for element related to an ElementAspect and retrieve Orgin, BBoxLow, BBoxHigh

### DISCUSSION THREAD

**Discussion by JohnMajerle** [iTwin/itwinjs-core#1788] [Q&A]
I bridged a dgn file that had ItemTypes applied to both graphic and parametric cells.

In the resulting iModel I find that these these items type definitions translate an ECClass named: **Signals.SignalElementAspect**.

I would like to determine the origin and range box of the geometric element related to my **_SignalElementAspect_**.  When I execute the following query form imodelconsole.bentley.com:

`select ere.ECClassId, s.Origin, s.BBoxLow, s.BBoxHigh, s.ECClassId, s.ECInstanceId, t.ECClassId, t.ECInstanceId, COUNT(*) from biscore.ElementOwnsMultiAspects ere JOIN BisCore.GeometricElement3d s ON s.ECInstanceId = ere.SourceECInstanceId JOIN Signals.SignalElementAspect t ON t.ECInstanceId = ere.TargetECInstanceId GROUP BY s.ECClassId, t.ECClassId`

... I get the following result:

![image](https://user-images.githubusercontent.com/30668739/124330948-1ae6c800-db54-11eb-9f12-67bd550ca412.png)

As you can see I get NULL for Origin, BBoxLow, and BBoxHigh.

What is the correct query that I should have used to get non-NULL values?

Thanks.

**Comment by bbastings** [comment]
User defined cells from DGN usually become "assemblies", that sounds like the case here. The cell components will become individual GeometricElement3ds with the cell header as their parent id. Since the cell header itself had no geometry, it's GeometricElement3d won't either, it's strictly for grouping purposes.

You could get the placement and element aligned box for each child element by querying for elements with the specified parent id...not sure that will suit your purposes.

To get the range of the assembly, you need to union the child ranges. That involves transforming each child's bbox values (element aligned box) by it's placement transform to get the axis aligned box for the child and using it to accumulate an overall range for all children. I don't think you can do this from imodelconsole...or at least I don't know how...

**Reply by pmconne** [reply]
This can be done using [built-in ECSql geometry functions](https://www.itwinjs.org/learning/geometrysqlfuncs/#imodel_bbox_union).

**Reply by JohnMajerle** [reply]
Hey Paul,
I'm having some trouble understanding the geometry function link you sent me.  The code snippet appears to have been taken from:
`imodeljs\example-code\snippets\src\backend\ECSQL-spatial-queries.test.ts`

I have a couple of questions.  First, my goal is to get this information from the frontend api, not the backend.  From the frontend the IModelConnection class doesn't have the method:
`iModel.models.getSubModel`  

Second, I'm not exactly sure what a **_sub model_** is and what this code snippet was intended to do.  It looks like it's getting the entire range of a model.  Going back to my example, I am able to query for the source instance related to my target ItemType instance.  Brian said that I need to accumulate the ranges of all the children of the source instance.  Does that mean that my source instance is actually a sub model?  If so, how do I get this sub model id from the frontend api?

Or am I totally misunderstanding how this is supposed to work?

**Reply by pmconne** [reply]
You can query from frontend or backend.
Ignore the details of the example; tweak the ECSql to select the element IDs of interest to you - probably using parent-child relationship (e.g., `select ECInstanceId from bis.GeometricElement3d where Parent.Id=YourAssemblyId`).

**Reply by JohnMajerle** [reply]
Hey Paul,

From imodelconsole I verified that my parent element has 11 children using:
`select ECInstanceId from bis.GeometricElement3d where Parent.Id=0x2000000465d`

I then modified my geometry query as follows:
```
SELECT
        iModel_bbox_union(
          iModel_placement_aabb(
            iModel_placement(
              iModel_point(child.Origin.X, child.Origin.Y, child.Origin.Z),
              iModel_angles(child.Yaw, child.Pitch, child.Roll),
              iModel_bbox(child.BBoxLow.X, child.BBoxLow.Y, child.BBoxLow.Z, child.BBoxHigh.X, child.BBoxHigh.Y, child.BBoxHigh.Z)
            )
          )
        )
      FROM  bis.GeometricElement3d AS child
        WHERE Parent.Id=0x2000000465d 
```
This displays the following result:

`1632247128711873965781891161611975306525216924121077986419121702222073187396572161492292035306525216924121077986463`

Ouch!  I guess this explains the sample code that takes this result and passes it into a call to:
`Range3d.fromArrayBuffer()`

Is there a way I can do the same thing from imodelconsole?  _(As much as possible I like to debug my ECSql in imodelconsole before using in an an application where its harder to debug.)_

Thanks

**Reply by pmconne** [reply]
> Is there a way I can do the same thing from imodelconsole?

Yes, use [iModel_bbox_value](https://www.itwinjs.org/learning/geometrysqlfuncs/#imodel_bbox_value) to extract the properties of the blob.

**Reply by JohnMajerle** [reply]
Hey Paul,

I'm having trouble querying for **_Range3d_** from the **_frontend_**.  See **_My Function_**, below.

During my test **_rangeSum_** successfully returns an **_AsyncIterableIterator_**.  But I don't know how to extract the **_ArrayBuffer_** from it to pass to Range3d.fromArrayBuffer().

Thanks,

John M.


**My Function**

    public static async GetRange (iModel: IModelConnection, parentId: Id64String) : Promise<Range3d> {
      const query = `
          SELECT
          iModel_bbox_union(
            iModel_placement_aabb(
              iModel_placement(
                iModel_point(child.Origin.X, child.Origin.Y, child.Origin.Z),
                iModel_angles(child.Yaw, child.Pitch, child.Roll),
                iModel_bbox(child.BBoxLow.X, child.BBoxLow.Y, child.BBoxLow.Z, child.BBoxHigh.X, child.BBoxHigh.Y, child.BBoxHigh.Z)
              )
            )
          )
          FROM  bis.GeometricElement3d AS child
              WHERE Parent.Id=${parentId} 
              `;

      const rangeSum = await iModel.query(query);          

      for await (const row of rangeSum)
          {
          // Below is wrong
          return Range3d.fromArrayBuffer(row);
          }  
      
      // Should not get here
      return Range3d.createNull();
      };
